File: python/composition/rpg.py
#----------------------------------------------------------------------|
class Character():
    def __init__(self, name, race, health, attack):
        self.name = name
        self.race = race
        self.copper = 0
        self.inv = Inventory([], 0, 0, 0)
        self.attack = attack
        self.health = health

    def battle(self, other):
        print(f"{self.name} attacks {other.name}")
        # self.health -= other.attack
        # other.health -= self.attack
        # if self.health <= 0:
        #     print(f"{self.name} was killed by {other.name}")

class Ranger(Character):
    def battle(self, other):
        print(f"{self.name} throws a sword at {other.name}")

class Thief(Character):
    def battle(self, other):
        print(f"{self.name} stabs {other.name} like a beardless wuss")

class Mage(Character):
    def __init__(self, name, race, health, attack):
        # super Class 
        super().__init__(name, race, health, attack)
        self.mana = 100

    def battle(self, other):
        print(f"{self.name} attacks {other.name}")
        mana -= 20

class Wizard(Character):
    def battle(self, other):
        print(f"{self.name} blasts {other.name}")



class Chest():
    def __init__(self, items, gold, silver, copper):
        self.inv = Inventory(items, gold, silver, copper)
        
class Inventory():
    def __init__(self, items, gold, silver, copper):
        self.items = items
        self.set_currency(gold, silver, copper) #delegation

    def transfer(self, to_inv):
        # Add items from self to to_inv
        to_inv.items.extend(self.items)
        # Delete items from self
        self.items = []
        # Add self copper to to_inv copper
        to_inv.copper += self.copper
        # Delete self copper
        self.copper = 0


    # This is a "Setter" - setters set attribute values
    def set_currency(self, gold,silver,copper):
        self.copper = gold *10000 + silver *100 + copper


    # A Getter method returns the value of atrributes - returns the 
    # values needed in the format needed
    def get_currency(self):
        gold = self.copper // 10000
        silver = (self.copper % 10000) // 100
        copper = self.copper % 100
        return (gold, silver, copper)


        

File: javascript/intro/rpg.js
// Character class definitions
class Character{
    constructor(name, race, health, attack){
        this.name = name
        this.race = race
        this.health = health
        this.attack = attack
        this.inv = new Inventory([], 0, 0, 0)
    }
    battle(other){
        console.log(`${this.name} attacks ${other.name}!`)
    }
}
class Ranger extends Character {
    battle(other){
        console.log(`${this.name} launches a brutal melee attack on ${other.name}!`)
    }
}
class Mage extends Character {
    constructor(name, race, health, attack){
        // Call the superclass constructor
        super(name, race, health, attack)
        this.mana = 100
    }
    battle(other){
        console.log(`${this.name} casts a wicked spell at ${other.name}!`)
        this.mana -= 20
    }
}
class Burglar extends Character{
    battle(other){
    console,log(`${this.name} sneaks in a stealth attack on ${other.name}!`)
    }
}

class Wizard extends Character{
    battle(other){
        console.log(`${this.name} summons an orc minion, which attacks ${other.name}!`)
    }
}
class Chest{
    constructor(items, gold, silver, copper){
        this.inv = new Inventory(items, gold, silver, copper)
    }
}
class Inventory{
    constructor(items, gold, silver, copper){
        this.items = items // list
        this.set_currency(gold, silver, copper) // Delegation
    }
    // from_inv and to_inv are instances of Inventory
    transfer(to_inv){
        // Add all the items from from_inv to to_inv
        to_inv.items.push(this.items)
        // # Delete all the items from from_inv
        this.items = []
        // # Add the currency from from_inv to to_inv
        to_inv.copper += this.copper
        // # Set currency of from_inv to 0
        this.copper = 0
    }
    // Setter
    set_currency(gold, silver, copper){
        this.copper = gold * 10000 + silver * 100 + copper
    }
    // # Getter
    get_currency(){
        let gold = Math.floor(this.copper / 10000)
        let silver = Math.floor((this.copper % 10000) / 100)
        let copper = this.copper % 100
        return [gold, silver, copper]
    }
}

let aragorn = new Ranger('Aragorn', 'Human', 100, 50)
let galadriel = new Mage('Galadriel', 'Elf', 120, 75)
let frodo = new Burglar('Frodo', 'Hobbit', 50, 25)
let saruman = new Wizard('Saruman', 'Human', 80, 100)

frodo.inv.set_currency(9, 47, 23)

chest = new Chest(['longsword', 'iron helm'], 2, 25, 99)

galadriel.battle(aragorn)

console.log(chest.inv)
console.log(aragorn)
console.log(frodo)
console.log(galadriel)

// # Frodo loots a chest!
console.log(frodo.inv)
chest.inv.transfer(frodo.inv)
console.log(frodo.inv)
console.log(frodo.inv.get_currency())
